# COGS 108 - EDA Checkpoint

## Authors

- Candice Tang: Conceptualization, Background Research
- Carter Cameron: Background Research, Writing - review & editing
- Owen Canoy: Analysis, Software, Writing - original draft
- Reece Capers: Ethics
- Yunjia Xing: Data Curation - original draft

# Research Question

How does the state-level median household income affect the academic performance (numerical average grades) for 8th grade students across the United States during 2018 - 2019 (before COVID-19 pandemic)?

## Background and Prior Work

Disparities in academic performance amongst students have long been attributed to differences in socioeconomic status among American students. Since the transition from middle school to high school marks the first step to adulthood, we want to look at the the academic performance for 8th graders and compare it to the socioeconmic status of their household. It is most often believed that students that perfrom well during 8th grade, tend to continue to perform well in highschool and further. Seeing as 8th grade has public testing and more accessible data, we will choose to analyze the relationship between socioeconomic status and academic performance within 8th graders.

A similar yet older study by The American Physiological Association analyzed the relationship between socioeconomic status and education. This study deemed, “Individuals within the top family income quartile are 8 times more likely to obtain a bachelor’s degree by age 24 as compared to individuals from the lowest family income quartile”2.

According to the National Assessment of Educational Progress (NAEP), 9 years old and 13 years old students in the United States have an increase in average scores for NAEP long-term trend (LTT) reading and mathematics assessments from 2018-2019 3. Meanwhile, with the rise of academic performance, the data from the United States Census Bureau has informed us that there is a 6.8 percent increase (from 64,323 to 68,703 dollars) in household income in the United States 7.

Maria N. Wodarz, Natalia L. Komarova and Timmy Ma had done research for the relationship of English Language Arts (ELA) and Mathematics performance scores in 3rd and 8th grades and the median household income from during the 2018-2019 school year. The research found out, in half of the states, there is a linear correlation between academic performance and household income, and the other half characterized by a saturating function. The results reveal that less affluent states exhibit a steeper increase in performance with income compared to wealthier states 1.

With the inquiry of the relationship between median household income and academic performance in 3rd and 8th grades, it is important to focus specifically on 8th grade students, as this stage represents a major academic and developmental transition. During 8th grade, students are preparing to enter high school, and their academic achievement can influence placement into advanced courses, access to academic resources, and long-term educational opportunities. By analyzing this relationship, this research can help identify whether economic differences between states are strongly associated with measurable gaps in academic performance.

^ Wodarz, Maria N., et al. “The Functional Form of the Association between K-12 Student Performance and Household Income in U.S. School Districts.” PLOS ONE, Public Library of Science, 10 Sept. 2025, journals.plos.org/plosone/article?id=10.1371%2Fjournal.pone.0329296.
^ “Education and Socioeconomic Status.” American Psychological Association, American Psychological Association, 2017, www.apa.org/pi/ses/resources/publications/education.
^ “NAEP Long-Term Trend Assessment Results:Reading and Mathematics.” The Nation’s Report Card, 2023, www.nationsreportcard.gov/ltt/?age=9.
^ Semega, Jessica. “Income and Poverty in the United States: 2019.” Census.Gov, 9 June 2022, www.census.gov/library/publications/2020/demo/p60-270.html.

# Hypothesis


There is a positive correlation between state-level median household income and academic performance (average grades) among U.S. 8th grade students during 2018-2019, such that states with higher median household income tend to have higher average student grades. We believe that higher household income means access to better academic resources and financial improvement shapes students self-confidence, therefore, with the increase of household income, students would have a better academic performance. We also believe that students that test better in 8th grade tend to have better academic success in high school.

## Data

In [5]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [12]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://stacks.stanford.edu/file/druid:cs829jn7849/seda_admindist_long_gcs_5.0_updated_20240319.csv', 'filename':'seda_admindist_long_gcs_5.0_updated_20240319.csv'},
    { 'url': 'https://stacks.stanford.edu/file/druid:cs829jn7849/seda_cov_admindist_long_5.0_updated_20240319.csv', 'filename':'seda_cov_admindist_long_5.0_updated_20240319.csv'}
]

# get_data.get_raw(datafiles,destination_directory='data/00-raw/')

### SEDA Long Admin Dataset:

This dataset holds the test scores that we will be caring about. The rest of this dataset is not too useful as we truly are just wanting to analyze the test scores in comparison to the socioeconomic status.

Bringing in imports to help identify, clean, tidy, and ultimately compare them.

In [13]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels
import scipy.stats as stats
import statsmodels.formula.api as smf
from statsmodels.stats.anova import anova_lm
import scipy.stats as stats
from scipy.stats import ttest_ind, chisquare, normaltest

We begin by loading the first dataset that contains the grade scores.

Let's take a quick look at what this dataset looks like and the column names we are dealing with:

In [14]:
test_score_long = 'data/00-raw/seda_admindist_long_gcs_5.0_updated_20240319.csv'
df_test_long = pd.read_csv(test_score_long)
df_long_shape = df_test_long.shape
df_long_colname = df_test_long.columns
print(f'This dataframe has {df_long_shape[0]:,} rows and {df_long_shape[1]} columns.')
display(df_test_long.head())
print(f'This dataframe has the following columns: {df_long_colname}')

This dataframe has 1,410,327 rows and 56 columns.


,fips,stateabb,sedaadmin,sedaadminname,subject,grade,year,gcs_mn_all,gcs_mn_se_all,tot_asmt_all,...,gcs_mn_whg,gcs_mn_se_whg,tot_asmt_whg,gcs_mn_wht,gcs_mn_se_wht,tot_asmt_wht,gcs_mn_wng,gcs_mn_se_wng,tot_asmt_wng,last_bie
0,1,AL,100002,Alabama Youth Services,mth,8,2009,2.683576,0.633632,36,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
1,1,AL,100002,Alabama Youth Services,rla,8,2009,3.399506,0.634232,29,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
2,1,AL,100002,Alabama Youth Services,mth,8,2011,2.553814,0.659750,27,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
3,1,AL,100002,Alabama Youth Services,rla,8,2011,5.077712,0.629885,29,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
4,1,AL,100002,Alabama Youth Services,mth,8,2012,3.533237,0.553637,31,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0


This dataframe has the following columns: Index(['fips', 'stateabb', 'sedaadmin', 'sedaadminname', 'subject', 'grade',
       'year', 'gcs_mn_all', 'gcs_mn_se_all', 'tot_asmt_all', 'gcs_mn_asn',
       'gcs_mn_se_asn', 'tot_asmt_asn', 'gcs_mn_blk', 'gcs_mn_se_blk',
       'tot_asmt_blk', 'gcs_mn_ecd', 'gcs_mn_se_ecd', 'tot_asmt_ecd',
       'gcs_mn_fem', 'gcs_mn_se_fem', 'tot_asmt_fem', 'gcs_mn_hsp',
       'gcs_mn_se_hsp', 'tot_asmt_hsp', 'gcs_mn_mal', 'gcs_mn_se_mal',
       'tot_asmt_mal', 'gcs_mn_mfg', 'gcs_mn_se_mfg', 'tot_asmt_mfg',
       'gcs_mn_nam', 'gcs_mn_se_nam', 'tot_asmt_nam', 'gcs_mn_nec',
       'gcs_mn_se_nec', 'tot_asmt_nec', 'gcs_mn_neg', 'gcs_mn_se_neg',
       'tot_asmt_neg', 'gcs_mn_wag', 'gcs_mn_se_wag', 'tot_asmt_wag',
       'gcs_mn_wbg', 'gcs_mn_se_wbg', 'tot_asmt_wbg', 'gcs_mn_whg',
       'gcs_mn_se_whg', 'tot_asmt_whg', 'gcs_mn_wht', 'gcs_mn_se_wht',
       'tot_asmt_wht', 'gcs_mn_wng', 'gcs_mn_se_wng', 'tot_asmt_wng',
       'last_bie'],
      dtype='obje

Now we have a slight idea of what we are working with, let's begin by shortening this extremely large dataframe. Since we are testing socioeconomic status and it's affect on test scores pre-pandemic, let's start by filtering out the data we don't need.

Since there is indeed a ton of data within this, it is easier to simply state the columns that we do care about and then clean up their column names once we have that situated. At the moment, we will be stuck with our slightly exotic column names since the majority of them will be dropped anyway. For now, bear with us.

We are going to filter our data to hold only the student information that we need. So, this will be filtering this data to hold only the sudents who are in 8th grade and under the year of 2019. We are choosing the year of 2019 since the 'year' column marks the spring of that year they took it. So, year of 2019 means the student took the test in the Spring of 2019.

In [15]:
# Create a mask that only holds True values for those that fit within
# our year and grade
grade_year_mask = (df_test_long['grade'] == 8 ) & (df_test_long['year'] == 2019)

# apply the mask to our data
df_grade_year_long = df_test_long[grade_year_mask]

# let's subset our data now to only hold the information we ultimately care about
# within this dataset
filtered_long = df_grade_year_long[['stateabb', 'sedaadmin', 'subject', 'gcs_mn_all']]

# take a quick peak at our new data
filtered_long_shape = filtered_long.shape
print(f'This dataframe has {filtered_long_shape[0]:,} rows and {filtered_long_shape[1]} columns.')
display(filtered_long.head())

This dataframe has 18,380 rows and 4 columns.


,stateabb,sedaadmin,subject,gcs_mn_all
135,AL,100005,mth,7.165455
136,AL,100005,rla,6.591630
267,AL,100006,mth,5.910346
268,AL,100006,rla,6.294259
397,AL,100007,mth,9.262524


We have already dramatically reduced the number of rows and columns. A quick note, gcs_mn_all represents the Administrative District Subject-Grade Year Mean for all students. In other words, this column is the test score column. This is our "success" metric in terms of academic success. We will explore later on what is deemed a good score.

Now that we have this taken care of for the first dataset, let's do a similar process for the next dataset.

### SEDA Covariate Dataset:
This set holds the data that will be tested in regards to socioeconomic status. By combining this dataset with the first one, we will be able to compare test scores in regards to socioeconomic status.

In [16]:
test_socio_cov = 'data/00-raw/seda_cov_admindist_long_5.0_updated_20240319.csv'
df_test_cov = pd.read_csv(test_socio_cov, low_memory=False)
df_cov_shape = df_test_cov.shape
df_cov_colname = df_test_cov.columns
print(f'This dataframe has {df_cov_shape[0]:,} rows and {df_cov_shape[1]} columns.')
display(df_test_cov.head())
print(f'This dataframe has the following columns: {df_cov_colname}')

This dataframe has 1,025,468 rows and 99 columns.


,sedaadmin,sedaadminname,year,grade,fips,stateabb,gslo,gshi,urban,suburb,...,single_momasn,single_momblk,single_momhsp,single_momnam,single_momwht,seswhtblk,seswhthsp,seswhtasn,seswhtnam,urbanicity
0,103430,Vestavia Hills City,2013,4.0,1,AL,Pre-Kindergarten,12,0.0,1.000000,...,0.046671,0.421355,0.168808,NaN,0.105810,2.561777,0.885265,-0.453838,NaN,Suburb
1,103150,Talladega City,2011,3.0,1,AL,Pre-Kindergarten,12,0.0,0.000000,...,NaN,0.541491,NaN,NaN,0.166315,3.034499,NaN,NaN,NaN,Rural
2,102730,Pickens County,2012,5.0,1,AL,Pre-Kindergarten,12,0.0,0.000000,...,NaN,0.565051,NaN,NaN,0.093022,4.715712,NaN,NaN,NaN,Rural
3,101730,Hartselle City,2018,3.0,1,AL,Pre-Kindergarten,12,0.0,0.546791,...,NaN,NaN,NaN,NaN,0.115748,NaN,NaN,NaN,NaN,Suburb
4,100480,Bullock County,2011,3.0,1,AL,Kindergarten,12,0.0,0.000000,...,NaN,0.494687,NaN,NaN,0.139562,3.106981,NaN,NaN,NaN,Town


This dataframe has the following columns: Index(['sedaadmin', 'sedaadminname', 'year', 'grade', 'fips', 'stateabb',
       'gslo', 'gshi', 'urban', 'suburb', 'town', 'rural', 'locale_city_large',
       'locale_city_midsize', 'locale_city_small', 'locale_suburb_large',
       'locale_suburb_midsize', 'locale_suburb_small', 'locale_town_fringe',
       'locale_town_distant', 'locale_town_remote', 'locale_rural_fringe',
       'locale_rural_distant', 'locale_rural_remote', 'pernam', 'perasn',
       'perhsp', 'perblk', 'perwht', 'perfl', 'perrl', 'perecd', 'perell',
       'perspeced', 'totenrl', 'diffexpecd_blkwht', 'diffexpecd_hspwht',
       'diffexpecd_namwht', 'diffexpecd_asnwht', 'diffexpmin2_blkwht',
       'diffexpmin2_hspwht', 'diffexpmin2_namwht', 'diffexpmin2_asnwht',
       'hswhtblk', 'hswhthsp', 'hsflnfl', 'hsecdnec', 'rswhtblk', 'rswhthsp',
       'rsflnfl', 'rsecdnec', 'perfrl', 'sesall', 'sesasn', 'sesblk', 'seshsp',
       'sesnam', 'seswht', 'lninc50all', 'lninc50asn',

Looking at the raw data of this file, we find that is has about 400,000 less rows but about 55 more columns. This indicates that we are trading off some of the rows for a greater set of socioeconomic indicators. Which is ideal for what we plan on look for. Like before we do have a lot of data that we wont use, albeit quite unique to consider. I will breif what we see above though, we see the different poverty rates, median incomes, proportion of students broken down by town/suburb/rural, amongst many other categories all splitting the percentages up between the ethnicty of the child and their family. There is plenty for us to work off of here, but we plan on keeping it relatively simple.

We are again going to filter down this data to only hold the information we need.

In [17]:
# Create a mask that only holds True values for those that fit within
# our year and grade
grade_year_covmask = (df_test_cov['grade'] == 8 ) & (df_test_cov['year'] == 2019)

# apply the mask to our data
df_grade_year_cov = df_test_cov[grade_year_covmask]

# instead of removing a lot of the columns we have, we are choosing to
# keep most columns for now until we properly determine what we should
# use as the true indicator of socioeconomic status
df_grade_year_cov = df_grade_year_cov.drop(columns=['gslo', 'gshi', 'fips', 'stateabb'])

# take a quick peak at our new data
filtered_cov_shape = df_grade_year_cov.shape
print(f'This dataframe has {filtered_cov_shape[0]:,} rows and {filtered_cov_shape[1]} columns.')
display(df_grade_year_cov.head())

This dataframe has 15,598 rows and 95 columns.


,sedaadmin,sedaadminname,year,grade,urban,suburb,town,rural,locale_city_large,locale_city_midsize,...,single_momasn,single_momblk,single_momhsp,single_momnam,single_momwht,seswhtblk,seswhthsp,seswhtasn,seswhtnam,urbanicity
10,102250,Marengo County,2019,8.0,0.000000,0.0,0.000000,1.000000,0.0,0.0,...,NaN,0.416911,NaN,NaN,0.041996,2.515431,NaN,NaN,NaN,Rural
11,101350,Escambia County,2019,8.0,0.000000,0.0,0.618328,0.381672,0.0,0.0,...,NaN,0.447900,NaN,0.155865,0.153912,2.850713,NaN,NaN,0.286388,Town
133,101530,Florence City,2019,8.0,1.000000,0.0,0.000000,0.000000,0.0,0.0,...,NaN,0.506204,NaN,NaN,0.181979,2.282934,NaN,NaN,NaN,City
187,100007,Hoover City,2019,8.0,0.926892,0.0,0.000000,0.073108,0.0,0.0,...,0.065341,0.363681,0.085382,NaN,0.097089,1.550844,1.615852,-0.057889,NaN,City
247,103150,Talladega City,2019,8.0,0.000000,0.0,0.259727,0.740273,0.0,0.0,...,NaN,0.677343,NaN,NaN,0.165913,2.729950,NaN,NaN,NaN,Rural


Looking at our filtered dataset, we do find a similar amount of filtered rows leftover. This is a good sign in retaining our useful data and ideally, we will soon find that our overall dataframe should not lose to many rows.

To merge the two datasets, we plan on doing as so by the 'sedaadmin' column. This column represents the SEDA Administrative District ID that gets used in nearly every school across the country. This should allow us to align all our data with the correct school district using the SEDA ID numbers as keys.

## Merging Our Data:

In [18]:
# save our two filtered df's from above
df_long = filtered_long
df_cov = df_grade_year_cov

# inact the .merger() to combine the two
df_combined = pd.merge(df_long, df_cov, on='sedaadmin', how='inner')

# display combined dataframe
combined_shape = df_combined.shape
combined_col = df_combined.columns
print(f'This dataframe has {combined_shape[0]:,} rows and {combined_shape[1]} columns.')
display(df_combined.head())
print(f'This dataframe has the following columns: {combined_col}')

This dataframe has 18,372 rows and 98 columns.


,stateabb,sedaadmin,subject,gcs_mn_all,sedaadminname,year,grade,urban,suburb,town,...,single_momasn,single_momblk,single_momhsp,single_momnam,single_momwht,seswhtblk,seswhthsp,seswhtasn,seswhtnam,urbanicity
0,AL,100005,mth,7.165455,Albertville City,2019,8.0,0.000000,0.0,1.0,...,NaN,NaN,0.148979,NaN,0.190042,NaN,1.551608,NaN,NaN,Town
1,AL,100005,rla,6.591630,Albertville City,2019,8.0,0.000000,0.0,1.0,...,NaN,NaN,0.148979,NaN,0.190042,NaN,1.551608,NaN,NaN,Town
2,AL,100006,mth,5.910346,Marshall County,2019,8.0,0.000000,0.0,0.0,...,NaN,NaN,0.218229,NaN,0.142449,NaN,1.309727,NaN,NaN,Rural
3,AL,100006,rla,6.294259,Marshall County,2019,8.0,0.000000,0.0,0.0,...,NaN,NaN,0.218229,NaN,0.142449,NaN,1.309727,NaN,NaN,Rural
4,AL,100007,mth,9.262524,Hoover City,2019,8.0,0.926892,0.0,0.0,...,0.065341,0.363681,0.085382,NaN,0.097089,1.550844,1.615852,-0.057889,NaN,City


This dataframe has the following columns: Index(['stateabb', 'sedaadmin', 'subject', 'gcs_mn_all', 'sedaadminname',
       'year', 'grade', 'urban', 'suburb', 'town', 'rural',
       'locale_city_large', 'locale_city_midsize', 'locale_city_small',
       'locale_suburb_large', 'locale_suburb_midsize', 'locale_suburb_small',
       'locale_town_fringe', 'locale_town_distant', 'locale_town_remote',
       'locale_rural_fringe', 'locale_rural_distant', 'locale_rural_remote',
       'pernam', 'perasn', 'perhsp', 'perblk', 'perwht', 'perfl', 'perrl',
       'perecd', 'perell', 'perspeced', 'totenrl', 'diffexpecd_blkwht',
       'diffexpecd_hspwht', 'diffexpecd_namwht', 'diffexpecd_asnwht',
       'diffexpmin2_blkwht', 'diffexpmin2_hspwht', 'diffexpmin2_namwht',
       'diffexpmin2_asnwht', 'hswhtblk', 'hswhthsp', 'hsflnfl', 'hsecdnec',
       'rswhtblk', 'rswhthsp', 'rsflnfl', 'rsecdnec', 'perfrl', 'sesall',
       'sesasn', 'sesblk', 'seshsp', 'sesnam', 'seswht', 'lninc50all',
       'lnin

Let's start to re-organize some of this data, to begin let's go through the columns that will certainly be used and give them a proper title.

In [30]:
# We are going to reduce our data to only fit all
# families. This way we can have a cleaner model
# and one that is more robust

# Removing all ethnicitcy specific columns
cols_to_keep = [
    'stateabb',
    'sedaadmin',
    'subject',
    'gcs_mn_all',
    'sedaadminname',
    'year',
    'grade',
    'urban',
    'suburb',
    'town',
    'rural',
    'locale_city_large',
    'locale_city_midsize',
    'locale_city_small',
    'locale_suburb_large',
    'locale_suburb_midsize',
    'locale_suburb_small',
    'locale_town_fringe',
    'locale_town_distant',
    'locale_town_remote',
    'locale_rural_fringe',
    'locale_rural_distant',
    'locale_rural_remote',
    'perfl',
    'perrl',
    'perecd',
    'totenrl',
    'perfrl',
    'lninc50all',
    'baplusall',
    'povertyall',
    'unempall',
    'snapall',
    'single_momall',
    'urbanicity'
]
# updated df with reduced columns
df_column = df_combined[cols_to_keep]

# renaming columns for ease of understanding
df_clean = df_column.rename(columns={
    'stateabb': 'State',
    'sedaadmin': 'District_ID',
    'sedaadminname': 'District_Name',
    'subject': 'Subject',
    'gcs_mn_all': 'Mean_Score_GCS',
    'year': 'Year',
    'grade': 'Grade',
    'urbanicity': 'Urbanicity_Index',
    'perfl': 'Percent_Free_Lunch',
    'perrl': 'Percent_Reduced_Lunch',
    'perecd': 'Percent_Econ_Disadv',
    'perfrl': 'Percent_FRL_Total',
    'totenrl': 'Total_Enrollment',
    'lninc50all': 'Median_Income_Log',
    'baplusall': 'Percent_BA_Degree',
    'povertyall': 'Poverty_Rate',
    'unempall': 'Unemployment_Rate',
    'snapall': 'Percent_SNAP',
    'single_momall': 'Percent_Single_Moms',
    'urban': 'Is_Urban',
    'suburb': 'Is_Suburb',
    'town': 'Is_Town',
    'rural': 'Is_Rural',
    'locale_city_large': 'City_Large',
    'locale_city_midsize': 'City_Midsize',
    'locale_city_small': 'City_Small',
    'locale_suburb_large': 'Suburb_Large',
    'locale_suburb_midsize': 'Suburb_Midsize',
    'locale_suburb_small': 'Suburb_Small',
    'locale_town_fringe': 'Town_Fringe',
    'locale_town_distant': 'Town_Distant',
    'locale_town_remote': 'Town_Remote',
    'locale_rural_fringe': 'Rural_Fringe',
    'locale_rural_distant': 'Rural_Distant',
    'locale_rural_remote': 'Rural_Remote'
})

# Let's take another look at this dataset
clean_shape = df_clean.shape
clean_col = df_clean.columns
print(f'This dataframe has {clean_shape[0]:,} rows and {clean_shape[1]} columns.')
display(df_clean.head())
print(f'This dataframe has the following columns: {clean_col}')

This dataframe has 18,372 rows and 35 columns.


,State,District_ID,Subject,Mean_Score_GCS,District_Name,Year,Grade,Is_Urban,Is_Suburb,Is_Town,...,Percent_Econ_Disadv,Total_Enrollment,Percent_FRL_Total,Median_Income_Log,Percent_BA_Degree,Poverty_Rate,Unemployment_Rate,Percent_SNAP,Percent_Single_Moms,Urbanicity_Index
0,AL,100005,mth,7.165455,Albertville City,2019,8.0,0.000000,0.0,1.0,...,0.384236,403.0,0.557839,10.640275,0.160546,0.234908,0.022281,0.169494,0.187094,Town
1,AL,100005,rla,6.591630,Albertville City,2019,8.0,0.000000,0.0,1.0,...,0.384236,403.0,0.557839,10.640275,0.160546,0.234908,0.022281,0.169494,0.187094,Town
2,AL,100006,mth,5.910346,Marshall County,2019,8.0,0.000000,0.0,0.0,...,0.683983,463.0,0.724273,10.665498,0.179877,0.187809,0.053178,0.108017,0.149927,Rural
3,AL,100006,rla,6.294259,Marshall County,2019,8.0,0.000000,0.0,0.0,...,0.683983,463.0,0.724273,10.665498,0.179877,0.187809,0.053178,0.108017,0.149927,Rural
4,AL,100007,mth,9.262524,Hoover City,2019,8.0,0.926892,0.0,0.0,...,0.243470,1076.0,0.257734,11.285567,0.572723,0.064096,0.034230,0.048099,0.140106,City


This dataframe has the following columns: Index(['State', 'District_ID', 'Subject', 'Mean_Score_GCS', 'District_Name',
       'Year', 'Grade', 'Is_Urban', 'Is_Suburb', 'Is_Town', 'Is_Rural',
       'City_Large', 'City_Midsize', 'City_Small', 'Suburb_Large',
       'Suburb_Midsize', 'Suburb_Small', 'Town_Fringe', 'Town_Distant',
       'Town_Remote', 'Rural_Fringe', 'Rural_Distant', 'Rural_Remote',
       'Percent_Free_Lunch', 'Percent_Reduced_Lunch', 'Percent_Econ_Disadv',
       'Total_Enrollment', 'Percent_FRL_Total', 'Median_Income_Log',
       'Percent_BA_Degree', 'Poverty_Rate', 'Unemployment_Rate',
       'Percent_SNAP', 'Percent_Single_Moms', 'Urbanicity_Index'],
      dtype='object')


#### Missing Data
Now to ensure that our combined dataset is fully wrangled, cleaned, and tidied. Let us clear up any of the remaining NaN values that we have so the we can work with our dataset.

Let's take a look at how much missing data we have:

In [20]:
missing_values = df_clean.isnull().sum()
missing_percent = (missing_values / len(df_clean)) * 100

missing_df = pd.DataFrame({'Missing Count': missing_values})

missing_df = missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False)
display(missing_df)

,Missing Count
Median_Income_Log,1140
Percent_BA_Degree,1140
Poverty_Rate,1140
Unemployment_Rate,1140
Percent_SNAP,1140
Percent_Single_Moms,1140
Percent_Econ_Disadv,156
Percent_Free_Lunch,7
Percent_Reduced_Lunch,7
Percent_FRL_Total,7


From the dataframe above, we can see that:

1. Six socioeconomic columns (`Median_Income_Log`, `Percent_BA_Degree`, `Poverty_Rate`, `Unemployment_Rate`, `Percent_SNAP`, `Percent_Single_Moms`) have the same 1140 missing counts.
2. 'Percent_Econ_Disadv has 156 missing values.
3. Three columns (`Percent_Free_Lunch`, `Percent_Reduced_Lunch`, `Percent_FRL_Total`)

Since the six socioeconomic columns share the same set of missing rows, they should lack the same socioeconomic datapoints. Omitting 1140 individuals from our observed population due to empty datapoints may cause bias if they are coming from some specific school districts that do not collect parental information.

Therefore, We choose to fill these missing values with the median of each column, as this value is less affected by extreme outliers and more representative of the 'average American household'.

Filling the missing values with median:

In [25]:
df_clean_processed = df_clean.copy()

cols_with_missing = ['Median_Income_Log', 'Percent_BA_Degree', 'Poverty_Rate', \
                     'Unemployment_Rate', 'Percent_SNAP', 'Percent_Single_Moms', \
                     'Percent_Econ_Disadv', 'Percent_Free_Lunch', \
                     'Percent_Reduced_Lunch', 'Percent_FRL_Total']

for col in cols_with_missing:
    median_val = df_clean_processed[col].median()
    df_clean_processed[col] = df_clean_processed[col].fillna(median_val)

df_clean_processed.head()

,State,District_ID,Subject,Mean_Score_GCS,District_Name,Year,Grade,Is_Urban,Is_Suburb,Is_Town,...,Percent_Econ_Disadv,Total_Enrollment,Percent_FRL_Total,Median_Income_Log,Percent_BA_Degree,Poverty_Rate,Unemployment_Rate,Percent_SNAP,Percent_Single_Moms,Urbanicity_Index
0,AL,100005,mth,7.165455,Albertville City,2019,8.0,0.000000,0.0,1.0,...,0.384236,403.0,0.557839,10.640275,0.160546,0.234908,0.022281,0.169494,0.187094,Town
1,AL,100005,rla,6.591630,Albertville City,2019,8.0,0.000000,0.0,1.0,...,0.384236,403.0,0.557839,10.640275,0.160546,0.234908,0.022281,0.169494,0.187094,Town
2,AL,100006,mth,5.910346,Marshall County,2019,8.0,0.000000,0.0,0.0,...,0.683983,463.0,0.724273,10.665498,0.179877,0.187809,0.053178,0.108017,0.149927,Rural
3,AL,100006,rla,6.294259,Marshall County,2019,8.0,0.000000,0.0,0.0,...,0.683983,463.0,0.724273,10.665498,0.179877,0.187809,0.053178,0.108017,0.149927,Rural
4,AL,100007,mth,9.262524,Hoover City,2019,8.0,0.926892,0.0,0.0,...,0.243470,1076.0,0.257734,11.285567,0.572723,0.064096,0.034230,0.048099,0.140106,City


Verifying that there is no missing values in the dataframe:

In [29]:
print(f"Remaining missing values: {df_clean_processed.isnull().sum().sum()}")

Remaining missing values: 0


## Results

### Exploratory Data Analysis

Instructions: replace the words in this subsection with whatever words you need to setup and preview the EDA you're going to do.   

Please explicitly load the fully wrangled data you will use from `data/02-processed`.  This is a good idea rather than forcing people to re-run the data getting / wrangling cells above.  Sometimes it takes a long time to get / wrangle data compared to reloading the fixed up dataset.

Carry out whatever EDA you need to for your project in the code cells below.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

Please note that you should consider the use of python modules in your work.  Any code which gets called repeatedly should be modularized. So if you run the same pre-processing, analysis or visualiazation on different subsets of the data, then you should turn that into a function or class.  Put that function or class in a .py file that lives in `modules/`.  Import the module you made and use it to get your work done.  For reference see `get_raw()` which is inside `modules/get_data.py`. 



#### Section 1 of EDA

here lies information tbd

#### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [ ]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

## Ethics

Instructions: Keep the contents of this cell. For each item on the checklist
-  put an X there if you've considered the item
-  IF THE ITEM IS RELEVANT place a short paragraph after the checklist item discussing the issue.
  
Items on this checklist are meant to provoke discussion among good-faith actors who take their ethical responsibilities seriously. Your teams will document these discussions and decisions for posterity using this section.  You don't have to solve these problems, you just have to acknowledge any potential harm no matter how unlikely.

Here is a [list of real world examples](https://deon.drivendata.org/examples/) for each item in the checklist that can refer to.

[![Deon badge](https://img.shields.io/badge/ethics%20checklist-deon-brightgreen.svg?style=popout-square)](http://deon.drivendata.org/)

### A. Data Collection
 - [X] **A.1 Informed consent**: If there are human subjects, have they given informed consent, where subjects affirmatively opt-in and have a clear understanding of the data uses to which they consent?

> The datasets we’re analyzing are derived from data collected through the decennial US census. The census is made compulsory for American households through Title 13, and non-respondents may be subject to fines or short-term incarceration. However, the federal government is required to maintain strict confidentiality through differential privacy.

 - [ ] **A.2 Collection bias**: Have we considered sources of bias that could be introduced during data collection and survey design and taken steps to mitigate those?
 - [X] **A.3 Limit PII exposure**: Have we considered ways to minimize exposure of personally identifiable information (PII) for example through anonymization or not collecting information that isn't relevant for analysis?
 
 > Participant anonymity is protected for 72 years by differential privacy, where statistical noise is added to the data to interfere with attempts to reidentify respondents. It maintains enough randomness to obscure data origin without compromising the integrity of the collection. Census data is re-released without redaction of personal information after 72 years have passed.
 
 - [ ] **A.4 Downstream bias mitigation**: Have we considered ways to enable testing downstream results for biased outcomes (e.g., collecting data on protected group status like race or gender)?

### B. Data Storage
 - [ ] **B.1 Data security**: Do we have a plan to protect and secure data (e.g., encryption at rest and in transit, access controls on internal users and third parties, access logs, and up-to-date software)?
 - [X] **B.2 Right to be forgotten**: Do we have a mechanism through which an individual can request their personal information be removed?
 
 > There is no mechanism in place to get individuals’ data removed from the Census. But personal information such as names, addresses, and contact information are removed from datasets before their information is added to the national database.

 - [X] **B.3 Data retention plan**: Is there a schedule or plan to delete the data after it is no longer needed?
 
 > The federal government appears to keep census records indefinitely. After 72 years, the unredacted data is made publicly available and kept that way for research and genealogical documentation.

### C. Analysis
 - [X] **C.1 Missing perspectives**: Have we sought to address blindspots in the analysis through engagement with relevant stakeholders (e.g., checking assumptions and discussing implications with affected communities and subject matter experts)?
 
 >Each state sets its own standards of public education and curriculum requirements, so students in different states likely have different experiences. Also, a low reported household income or a poor credit score may not be an indicator of lived financial hardship. The data only allows us to speculate on the living situation rather than explicitly stating it.

 - [ ] **C.2 Dataset bias**: Have we examined the data for possible sources of bias and taken steps to mitigate or address these biases (e.g., stereotype perpetuation, confirmation bias, imbalanced classes, or omitted confounding variables)?
 - [ ] **C.3 Honest representation**: Are our visualizations, summary statistics, and reports designed to honestly represent the underlying data?
 - [ ] **C.4 Privacy in analysis**: Have we ensured that data with PII are not used or displayed unless necessary for the analysis?
 - [X] **C.5 Auditability**: Is the process of generating the analysis well documented and reproducible if we discover issues in the future?
 
 >The datasets used to generate this analysis are publicly available and accessible.

### D. Modeling
 - [X] **D.1 Proxy discrimination**: Have we ensured that the model does not rely on variables or proxies for variables that are unfairly discriminatory?
 
 >It can be argued that financial disparities in this country are inherently discriminatory, given the nation’s history. The data observed is not race-specific, but the races of respondents are included.

 - [X] **D.2 Fairness across groups**: Have we tested model results for fairness with respect to different affected groups (e.g., tested for disparate error rates)?
 
 >Studies of this kind have been conducted across various demographics and in several other countries around the world. Adjusting for the local cost of living in each state will result in a model that is fairly generalizable.

 - [ ] **D.3 Metric selection**: Have we considered the effects of optimizing for our defined metrics and considered additional metrics?
 - [X] **D.4 Explainability**: Can we explain in understandable terms a decision the model made in cases where a justification is needed?
 
 >The model is intended to identify possible correlations between a household’s degree of financial hardship and the children’s performance in school. 

 - [ ] **D.5 Communicate limitations**: Have we communicated the shortcomings, limitations, and biases of the model to relevant stakeholders in ways that can be generally understood?

### E. Deployment
 - [ ] **E.1 Monitoring and evaluation**: Do we have a clear plan to monitor the model and its impacts after it is deployed (e.g., performance monitoring, regular audit of sample predictions, human review of high-stakes decisions, reviewing downstream impacts of errors or low-confidence decisions, testing for concept drift)?
 - [X] **E.2 Redress**: Have we discussed with our organization a plan for response if users are harmed by the results (e.g., how does the data science team evaluate these cases and update analysis and models to prevent future harm)?
 
 >Users should not face retribution for displaying financial need or academic difficulties. This model is intended to identify need so financial support can be provided.

 - [ ] **E.3 Roll back**: Is there a way to turn off or roll back the model in production if necessary?
 - [ ] **E.4 Unintended use**: Have we taken steps to identify and prevent unintended uses and abuse of the model and do we have a plan to monitor these once the model is deployed?


## Team Expectations 

* We expect that our members will communicate accurately, frequently, and respectfully with one another in order to keep a good workflow and environment. We will use Discord as our main platform of communication, and expect that any unexpected leaves or holds with our members to be informed to the rest of our group.
* We expect that decisions will be made as a majority agreement with adjustments made to cater closer to everyone’s preferences. When some members are unavailable, the remaining members will decide on executive decisions as a smaller group.
* We expect that all members of the project will contribute equally on different parts of it. When on  e has trouble with getting something done, they will reach out to the rest of us so we can provide assistance in figuring out how to complete the task.  If a member falls behind or is seen slacking off on doing what they are supposed to do, we will notify them to pick up the pace and make sure to increase their productivity quickly.

## Project Timeline Proposal

| Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 2/4  |  5:30 PM | Read and think about the COGS108 project proposal and get an idea of what to do. | Decide what research question we should ask; Find proper dataset and sources to use. | 
| 2/18  | 5:30 PM  | Review all data and and properly access which data will certainly be used. Conclude data Checkpint file.| Discussion analyzation methods and clear up any confusion with data. |
| 2/25  | 6 PM  | Ensure >90% of data analysis is completed and start conducting a conclusion based on data at hand. | Reflect on what analysis suggested, discuss finalization steps and talk about video ideas/submission.  |
| 3/5  | 5:30 PM  |  Review the project as a whole and ensure everything looks correct. | Touch on any topics/conclusions that may be unclear and ensure everyone is satisfied with the project. |
| 3/12 | 5:00 PM | Adjust any last minute feedback and concerns within the group. | Discuss video/presentation understanding and ideas |